core/arch/arm/plat-synquacer/rng_pta.c

```diff
@@ -0,0 +1,211 @@
+{
+	uint32_t i, crc = 0xffffffff;
+
+	for (i = 0; i < sz; i++)
+		crc = (crc >> 8) ^ table[(crc ^ msg[i]) & 0xff];
+
+	return ~crc;
+}
+
+static void pool_add_entropy(uint8_t byte)
+{
+	entropy_pool[entropy_size] = byte;
+	entropy_size++;
+}
+
+static void pool_get_entropy(uint8_t *buf, uint32_t size)
+{
+	uint32_t off;
+
+	if (size > entropy_size)
+		return;
+
+	off = entropy_size - size;
+
+	memcpy(buf, &entropy_pool[off], size);
+	entropy_size -= size;
+}
+
+static uint32_t pool_get_entropy_sz(void)
+{
+	uint32_t sz;
+
+	sz = entropy_size;
+
+	return sz;
+}
+
+void rng_collect_entropy(void)
+{
+	uint64_t ts0 = 0, ts1 = 0, ts2 = 0, ts3 = 0, ts4 = 0, ts5 = 0, ts6 = 0;
+	uint8_t e = 0;
+
+	ts0 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE0 +
+					      TEMP_DATA_REG_OFFSET));
+	ts1 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE1 +
+					      TEMP_DATA_REG_OFFSET));
+	ts2 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE2 +
+					      TEMP_DATA_REG_OFFSET));
+	ts3 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE3 +
+					      TEMP_DATA_REG_OFFSET));
+	ts4 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE4 +
+					      TEMP_DATA_REG_OFFSET));
+	ts5 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE5 +
+					      TEMP_DATA_REG_OFFSET));
+	ts6 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE6 +
+					      TEMP_DATA_REG_OFFSET));
+
+	e = ((ts0 & 0x1) << 6) | ((ts1 & 0x1) << 5) | ((ts2 & 0x1) << 4) |
+	    ((ts3 & 0x1) << 3) | ((ts4 & 0x1) << 2) | ((ts5 & 0x1) << 1) |
+	    ((ts6 & 0x1) << 0);
+
+	e |= ((ts0 ^ ts1 ^ ts2 ^ ts3 ^ ts4 ^ ts5 ^ ts6) << 6) & 0x80;
+
+	pool_add_entropy(e);
+
+	if (pool_get_entropy_sz() >= ENTROPY_POOL_SIZE)
+		generic_timer_stop();
+}
+
+static TEE_Result rng_get_entropy(uint32_t types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint8_t *e = NULL;
+	uint32_t i, digest, pool_size = 0, rq_size = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	rq_size = params[0].memref.size;
+
+	if ((rq_size == 0) || (rq_size > ENTROPY_POOL_SIZE))
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	e = (uint8_t *)params[0].memref.buffer;
+	if (e == NULL)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	pool_size = pool_get_entropy_sz();
+
+	if (pool_size < rq_size) {
+		params[0].memref.size = pool_size;
+		pool_get_entropy(e, pool_size);
+	} else {
+		params[0].memref.size = rq_size;
+		pool_get_entropy(e, rq_size);
+	}
+
+	/* CRC whitening */
+	for (i = 0; i < (params[0].memref.size/4); i++) {
+		digest = crc32((e + 4*i), 4);
+		memcpy((e + 4*i), &digest, sizeof(uint32_t));
+	}
+
+	/* Enable timer FIQ to fetch entropy */
+	generic_timer_start();
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Trusted Application Entry Points
+ */
+static TEE_Result open_session(uint32_t nParamTypes __unused,
```

Comment:
We [try to follow Linux kernel coding style with one exception, that types, function names etc from the GP spec(s) are allowed to be written as specified in the GP spec(s)](https://github.com/OP-TEE/optee_os#5-coding-standards). Basically most the parameters here should be lower case and with underscore if more than one word.

Reply:
Ok will follow this.